fetchurl: Add hook for rewriting/filtering URLs

jherland opened the pull request:
This allows on-the-fly rewriting of URLs before they are passed from `fetchurl` (or `fetchurlBoot`) to `curl`.

The intended use is to allow inserting company-internal mirrors, or working around company firewalls and similar network restrictions, without having to extensively patch across all of nixpkgs. Instead, users can pass a function in their nixpkgs config that performs the necessary URL rewrites.

We first tried solving this with a shell hook called from inside `fetchurl`'s `builder.sh`, as that would apply rewrites _after_ the `mirror://` scheme has been resolved, however, this was not feasible to also apply to `fetchurlBoot`, hence we instead settled on a Nix function called before the builder is invoked. If we still want/need the rewrite to happen _after_ resolving `mirror://` URLs, then we probably need to move the mirroring logic out of the `builder.sh` and rewrite it in nix.

This is related in spirit to https://github.com/NixOS/nixpkgs/pull/409010, but does not directly depend on it.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] Added a release notes entry about the new config.rewriteURL attribute
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

toonn commented on pkgs/build-support/fetchurl/boot.nix:
```diff
@@ -28,7 +31,15 @@ import <nix/fetchurl.nix> {
+      m = builtins.match "mirror://([a-z]+)/(.*)" url_;
     in
-    if m == null then url else builtins.head (mirrors.${builtins.elemAt m 0}) + (builtins.elemAt m 1);
+    if m == null then url_ else builtins.head (mirrors.${builtins.elemAt m 0}) + (builtins.elemAt m 1);
```
So the problem is this rewriting can't happen in `builder.sh` after the resolving of the mirror URLs?

jherland replied:
Indeed. AFAICS, we must choose whether we want rewrites to happen before or after mirrors are resolved, and depending on that choice, we might have to rewrite larger parts of the fetchurl implementation.

jherland committed:
fetchurl: Add hook for rewriting/filtering URLs

This allows on-the-fly rewriting of URLs before they are passed from
fetchurl (or fetchurlBoot) to curl.

The intended use is to allow inserting company-internal mirrors, or
working around company firewalls and similar network restrictions,
without having to extensively patch across all of nixpkgs. Instead,
users can pass a function in their nixpkgs that performs the necessary
URL rewrites.

Co-authored-by: Alexander Bantyev <balsoft@balsoft.ru>

philiptaron commented on pkgs/top-level/config.nix:
```diff
@@ -211,6 +211,27 @@ let
+
+        The intended use is to allow URL rewriting to insert company-internal mirrors, or work around company firewalls and similar network restrictions.
+      '';
+      default = (url: url);
```
this is `lib.id`

philiptaron reviewed (approved):
I'm amenable to this feature. 

I do wonder if it's not going to be thorough enough for the stated usecase. Will you actually need to reach into the derivation and mutate (say) the environment variables also? I suspect you will, in order to set environment variables needed by proxies and the like.

If that's the case, you're going to be looking for a heavier hammer than this. We have those hammers! We can make that work!

jherland commented:
@philiptaron writes:
> I do wonder if it's not going to be thorough enough for the stated usecase. Will you actually need to reach into the derivation and mutate (say) the environment variables also? I suspect you will, in order to set environment variables needed by proxies and the like.

For proxies and similar we should already be able to pass those in via `impureEnvVars`, shouldn't we?

> If that's the case, you're going to be looking for a heavier hammer than this. We have those hammers! We can make that work!

For now, we're simply looking to do bulk overrides of URLs onto company-internal mirrors, and to do so without causing world rebuilds, or otherwise hacking each individual derivation. What are the hammers you're referring to, and how would those help us here?

philiptaron commented:
The hammers I'm thinking of equate to an overlay replacing fetchurl.

balsoft commented:
> The hammers I'm thinking of equate to an overlay replacing fetchurl.

I'd like to note that that wouldn't work for `fetchurlBoot`, while this change transparently works for both `fetchurl` and `fetchurlBoot`.

trofi commented:
Bisect says abda866f172173b6fd16ccec9b15fb1cbb649cb2 `fetchurl: Add hook for rewriting/filtering URLs` broke the eval of `minimal-bootstrap.bash` on `master` as:

```
$ nix-instantiate -A minimal-bootstrap.bash --argstr system i686-linux
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating derivation 'bash-5.2.15'
         whose name attribute is located at pkgs/os-specific/linux/minimal-bootstrap/bash/2.nix:85:19

       … while evaluating attribute 'PATH' of derivation 'bash-5.2.15'
         at pkgs/os-specific/linux/minimal-bootstrap/bash/2.nix:108:11:
          107|           SHELL = "${bash_2_05}/bin/bash";
          108|           PATH = lib.makeBinPath (
             |           ^
          109|             (env.nativeBuildInputs or [ ])

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: function 'anonymous lambda' called without required argument 'rewriteURL'
       at /tmp/nixpkgs/pkgs/build-support/fetchurl/boot.nix:5:1:
            4|
            5| {
             | ^
            6|   rewriteURL,
```

trofi commented:
Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/414769
